src/sentry/api/utils.py

```diff
@@ -32,13 +40,38 @@ def get_datetime_from_stats_period(stats_period, now=None):
         raise InvalidParams(f"Invalid statsPeriod: {stats_period!r}")
 
 
-def default_start_end_dates(now=None, default_stats_period=MAX_STATS_PERIOD):
+def default_start_end_dates(
+    now: datetime.datetime | None = None,
+    default_stats_period: datetime.timedelta = MAX_STATS_PERIOD,
+) -> tuple[datetime.datetime, datetime.datetime]:
     if now is None:
         now = timezone.now()
     return now - default_stats_period, now
 
 
-def get_date_range_from_params(params, optional=False, default_stats_period=MAX_STATS_PERIOD):
+@overload
+def get_date_range_from_params(
+    params: dict[str, Any],
+    optional: Literal[False] = ...,
+    default_stats_period: datetime.timedelta = ...,
+) -> tuple[datetime.datetime, datetime.datetime]:
+    ...
+
+
+@overload
+def get_date_range_from_params(
+    params: dict[str, Any],
+    optional: bool = ...,
+    default_stats_period: datetime.timedelta = ...,
+) -> tuple[None, None] | tuple[datetime.datetime, datetime.datetime]:
+    ...
+
+
+def get_date_range_from_params(
+    params: dict[str, Any],
+    optional: bool = False,
+    default_stats_period: datetime.timedelta = MAX_STATS_PERIOD,
+) -> tuple[None, None] | tuple[datetime.datetime, datetime.datetime]:
```

Comment:
Checking my understanding: 
The final, non-overload definition is ignored by the typechecker, yes?

Reply:
nope -- mypy combines the overloads with the implementation